File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.DynamicProxyExample/Implementations.cs
namespace ProxyDesignPattern.DynamicProxyExample;

/// <summary>
/// Real implementation of ICalculator.
/// </summary>
public class Calculator : ICalculator
{
    public int Add(int a, int b) => a + b;
    public int Subtract(int a, int b) => a - b;
    public int Multiply(int a, int b) => a * b;
    public double Divide(int a, int b) => b != 0 ? (double)a / b : throw new DivideByZeroException();
}

/// <summary>
/// Real implementation of IGreetingService.
/// </summary>
public class GreetingService : IGreetingService
{
    public string Greet(string name) => $"Hello, {name}!";
    public string Farewell(string name) => $"Goodbye, {name}!";
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.DynamicProxyExample/src/Implementations.ts
import { ICalculator, IGreetingService } from "./Interfaces";

/**
 * Real implementation of ICalculator.
 */
export class Calculator implements ICalculator {
  add(a: number, b: number): number {
    return a + b;
  }

  subtract(a: number, b: number): number {
    return a - b;
  }

  multiply(a: number, b: number): number {
    return a * b;
  }

  divide(a: number, b: number): number {
    if (b === 0) throw new Error("Division by zero");
    return a / b;
  }
}

/**
 * Real implementation of IGreetingService.
 */
export class GreetingService implements IGreetingService {
  greet(name: string): string {
    return `Hello, ${name}!`;
  }

  farewell(name: string): string {
    return `Goodbye, ${name}!`;
  }
}
